feat: cgt

tynes opened the pull request:
**Description**

Based on top of https://github.com/ethereum-optimism/optimism/pull/17320

Enables OP Stack chains to use an asset other than ETH as their native fee currency by introducing a minimal, flexible, and standardizable Custom Gas Token (CGT) architecture. To do so, it introduces new **predeploys** (`NativeAssetLiquidity` and `LiquidityController`) that decouple native asset management from core bridging operations while using a single `isCustomGasToken()` flag that disables ETH transfer flows in all bridging methods. The two new pre-deploys manage the native asset supply.

**Tests**

Core contracts:

- `OptimismPortal` and `L1Block`: Tests for `isCustomGasToken()` flag blocking ETH deposits/withdrawals
- `NativeAssetLiquidity`: Tests deposit/withdraw access control and functionality
- `LiquidityController`: Tests minter authorization, mint/burn operations, and metadata functions

**Metadata**

- Spec: https://github.com/ethereum-optimism/specs/pull/749
- Design Doc: https://github.com/ethereum-optimism/design-docs/pull/305

---------------------------------

### Contributors
  - @0xOneTony
  - @0xniha
  - @agusduha
  - @ashitakah
  - @hexshire
  - @Joxess
  - @tynes

hexshire committed:
feat(cgt): custom gas token

Signed-off-by: Hex <165055168+hexshire@users.noreply.github.com>
Co-authored-by: niha <205694301+0xniha@users.noreply.github.com>
Co-authored-by: Ashitaka <96790496+ashitakah@users.noreply.github.com>
Co-authored-by: AgusDuha <81362284+agusduha@users.noreply.github.com>

0xniha committed:
feat: add gasPayingTokenName and gasPayingTokenSymbol on json config files

0xniha committed:
fix: contracts semver

ashitakah committed:
fix: remove system config bool

0xniha committed:
test: add OptimismPortal2CGT tests

0xniha committed:
chore(cgt): set cgt flag l1block & fixes

hexshire committed:
fix(linter): resolve goimports formatting issue

0xniha committed:
feat: add separate l2 contracts for cgt (#530)

* feat: add L1BlockCGT

* feat: add L2ToL1MessagePasserCGT

* chore: remove test exclution in test validation

hexshire committed:
test(cgt): fix failing tests (#529)

* test(cgt): fix failing tests

* test(cgt): fix portal version

* test(cgt): skip tests on forked mode

* fix: predeploys test cgt mismatch

* test(cgt): minor fixes

---------

Co-authored-by: niha <205694301+0xniha@users.noreply.github.com>

hexshire committed:
fix(cgt): revert weth

tynes committed:
cgt: feature flag integration

Updates the implementation to use the new feature flagging system

tynes committed:
config: make backwards compatible

Use the legacy config name so that less needs to change for legacy CGT
chains

tynes committed:
fix: build issue

Remove leftover merge conflict

tynes committed:
fix: better backwards compatibility

tynes committed:
build: fix

tynes committed:
lint: fix

tynes committed:
snapshots: update

tynes committed:
op-deployer: apply test with CGT

Add a specific test for CGT in the intent

tynes committed:
contracts-bedrock: fix versioning

tynes committed:
lint: fix

tynes committed:
deployer: remove standard values

tynes committed:
contracts: semver lock

tynes committed:
contracts-bedrock: fix semver + abis

tynes committed:
cgt: solidity test cleanup

tynes committed:
solidity: fmt

tynes committed:
tests: remove dead imports

tynes committed:
tests: fixup

tynes commented:
Looks like I am getting a bunch of failures due to version mismatches. It works when not running in a forked environment, so I am guessing it is something to do with u16a not being deployed and upgraded to yet?

Run tests:
```
FOUNDRY_PROFILE=lite forge test --mc OPContractsManagerStandardValidator_PermissionedDisputeGame_Test
```

Example failure:
```
[FAIL: assertion failed: PDDG-40 != PORTAL-10,PDDG-40] test_validate_permissionedDisputeGameInvalidAbsolutePrestate_succeeds() (gas: 715767)
```

tynes commented on packages/contracts-bedrock/scripts/L2Genesis.s.sol:
```diff
@@ -236,6 +242,10 @@ contract L2Genesis is Script {
         }
+        if (_input.useCustomGasToken) {
+            setLiquidityController(_input); // 29
+            setNativeAssetLiquidity(); // 2A
```
Need to add a config here that allows for setting the value that is in the `NativeAssetLiquidity` contract so it can be set to 0

tynes committed:
cgt: configurable liquidity amount

tynes committed:
feat: configurable native asset liquidity balance

Fix the build for this feature

tynes committed:
cleanup: merge L1Block logic so that we inherit

Keeps it simple because development on Jovian is happening and we will
inherit any Jovian modifications automatically

tynes committed:
snapshots: update

tynes committed:
cgt: inherit logic for L2ToL1MessagePasser

Generally simplifies the code

tynes committed:
semver-lock: update

tynes committed:
lint: fix

tynes committed:
lint: fixup

tynes committed:
interfaces: fix

tynes committed:
fixes: smol

tynes committed:
deploy-config: sane default

tynes committed:
lint: fix

tynes committed:
linting: fix

tynes committed:
lint: fix

tynes committed:
semgrep: fix

tynes committed:
lint: fix

tynes commented on op-deployer/pkg/deployer/state/intent_test.go:
```diff
@@ -211,3 +237,11 @@ func setFeeAddresses(intent *Intent) {
 }
+
+func setCustomGasToken(intent *Intent) {
+	intent.Chains[0].CustomGasToken = &CustomGasToken{
```
Should make sure the native asset liquidity amount is set properly everywhere it needs to be

0xniha replied:
Updated here: [9bb50bd](https://github.com/ethereum-optimism/optimism/pull/17412/commits/9bb50bd10b1966f45895241bf7c37d4545d94946)

tynes committed:
tests: fix

tynes committed:
tests: fix fuzz

smartcontracts commented on packages/contracts-bedrock/src/L1/OptimismPortal2.sol:
```diff
@@ -259,6 +259,12 @@ contract OptimismPortal2 is Initializable, ResourceMetering, ReinitializableBase
 
+    /// @notice Returns whether the custom gas token feature is enabled.
+    /// @return bool True if the custom gas token feature is enabled, false otherwise.
+    function isCustomGasToken() public view returns (bool) {
```
Probably don't need this, can just have users use the SystemConfig

agusduha replied:
We can remove it

agusduha replied:
Fixed https://github.com/ethereum-optimism/optimism/pull/17412/commits/9bb50bd10b1966f45895241bf7c37d4545d94946

smartcontracts commented on packages/contracts-bedrock/test/L1/OptimismPortal2.t.sol:
```diff
@@ -274,56 +293,6 @@ contract OptimismPortal2_Initialize_Test is OptimismPortal2_TestInit {
-
-    /// @notice Tests that the initialize function reverts if called by a non-proxy admin or owner.
-    /// @param _sender The address of the sender to test.
-    function testFuzz_initialize_interopNotProxyAdminOrProxyAdminOwner_reverts(address _sender) public {
```
Why were these deleted?

agusduha replied:
These shouldn't be deleted, reverting

agusduha replied:
Fixed https://github.com/ethereum-optimism/optimism/pull/17412/commits/9bb50bd10b1966f45895241bf7c37d4545d94946

bitwiseguy commented on op-deployer/pkg/deployer/state/intent.go:
```diff
@@ -343,6 +349,12 @@ func NewIntentStandard(l1ChainId uint64, l2ChainIds []common.Hash) (Intent, erro
+				Name:                       "",
+				Symbol:                     "",
+				NativeAssetLiquidityAmount: nil,
+			},
```
Should we add a check in `func (c *Intent) validateStandardValues()` to ensure `CustomGasToken` is not used by "standard" chains? The definition of "standard" is murky in my mind, but I think it means that the chain _would_ be compatible with op-mainnet if they were part of the same interop set. If a chain has `CustomGasToken` enabled, would it violate that definition of "standard"?

agusduha replied:
It is true that the interop feature is not already compatible with CGT, but there is ways to do it that we are planning to attack in the future. I think that CGT can be considered standard once these are solved but will let @mds1 or @tynes to answer

0xniha replied:
Yes, the [`validateStandardValues()`](https://github.com/ethereum-optimism/optimism/blob/9bb50bd10b1966f45895241bf7c37d4545d94946/op-deployer/pkg/deployer/state/intent.go#L161) already has the check for cgt:
```ts
if chain.CustomGasToken.Enabled {
   return fmt.Errorf("%w: chainId=%s custom gas token must be nil for standard chains", ErrNonStandardValue, chain.ID)
}
```

bitwiseguy replied:
Ok great. Maybe change the error message to say "custom gas token must be disabled for standard chains"?

agusduha replied:
Fixed https://github.com/ethereum-optimism/optimism/pull/17412/commits/4ad8a301948a09f77c2d239ef15604d76fd8546c

bitwiseguy commented on packages/contracts-bedrock/scripts/deploy/DeployOPChain.s.sol:
```diff
@@ -107,13 +108,21 @@ contract DeployOPChainInput is BaseDeployIO {
+        if (_sel == this.disputeAbsolutePrestate.selector) {
+            _disputeAbsolutePrestate = Claim.wrap(_value);
+        } else {
+            revert("DeployImplementationsInput: unknown selector");
```
This revert message seems wrong (although the message was not introduced by this pr). Shouldn't it be `DeployOPChainInput: unknown selector`

agusduha replied:
Fixed https://github.com/ethereum-optimism/optimism/pull/17412/commits/dfe5282a958e2aeac3fce599e4a2f0cb174c6cb5

bitwiseguy commented on op-deployer/pkg/deployer/integration_test/apply_test.go:
```diff
@@ -238,6 +240,43 @@ func TestEndToEndApply(t *testing.T) {
+		}, nil)
+		require.NoError(t, err)
+
+		var response bool
+		err = fn.DecodeReturns(res, &response)
+		require.NoError(t, err)
+		require.Equal(t, true, response)
```
Are there any more in-depth L1 checks we can/should make to ensure custom gas token is properly configured besides just checking this boolean?

Maybe we could check the L2 genesis for some CGT-related fields as well for more completeness?

agusduha replied:
Agree, we will add checks in the L2 genesis script

0xniha replied:
We added a new check to assert that `nativeAssetLiquidityAmount` is correctly set in [`apply_test.go`](https://github.com/ethereum-optimism/optimism/blob/9bb50bd10b1966f45895241bf7c37d4545d94946/op-deployer/pkg/deployer/integration_test/apply_test.go#L294).

Could you expand on which checks you would like to add for the L2 genesis?

bitwiseguy commented on op-deployer/pkg/deployer/state/deploy_config.go:
```diff
@@ -71,6 +71,13 @@ func CombineDeployConfig(intent *Intent, chainIntent *ChainIntent, state *State,
 				EIP1559Elasticity:        chainIntent.Eip1559Elasticity,
 			},
 
+			GasTokenDeployConfig: genesis.GasTokenDeployConfig{
+				UseCustomGasToken:          chainIntent.CustomGasToken.Enabled,
+				GasPayingTokenName:         chainIntent.CustomGasToken.Name,
+				GasPayingTokenSymbol:       chainIntent.CustomGasToken.Symbol,
+				NativeAssetLiquidityAmount: chainIntent.CustomGasToken.NativeAssetLiquidityAmount,
+			},
```
Do we need to check for a nil `chainIntent.CustomGasToken` here since its a pointer? Should we make `chainIntent.CustomGasToken` a non-pointer to avoid the extra logic needed to check for nil?

agusduha replied:
Agree, we should be using the same `GetNativeAssetLiquidityAmount()` that we used in the l2genesis.go

0xniha replied:
Cool, refactored here: [9bb50bd](https://github.com/ethereum-optimism/optimism/pull/17412/commits/9bb50bd10b1966f45895241bf7c37d4545d94946)

bitwiseguy commented on op-deployer/pkg/deployer/state/chain_intent.go:
```diff
@@ -56,6 +57,13 @@ type L2DevGenesisParams struct {
+	Enabled                    bool         `json:"enabled" toml:"enabled"`
+	Name                       string       `json:"name" toml:"name"`
+	Symbol                     string       `json:"symbol" toml:"symbol"`
+	NativeAssetLiquidityAmount *hexutil.Big `json:"nativeAssetLiquidityAmount,omitempty" toml:"nativeAssetLiquidityAmount,omitempty"`
```
Are all of these terms equivalent? If so could we choose one and use it consistently? Since the feature is "cgt" seems like "CustomGasToken" should be the term we use.

* NativeAsset
* CustomGasToken
* GasPayingToken

agusduha replied:
Agree that the namings are not consistent, probably CustomGasToken or NativeAsset are suitable for this (in particular for name and symbol). I suggest opening an issue to track it to reduce the diff in this PR if that's ok

bitwiseguy replied:
Could we change the one chainIntent field name from `CustomGasToken.NativeAssetLiquidityAmount` to `CustomGasToken.InitialAmount` or `CustomGasToken.GenesisAmount` (assuming thats what the field represents)? The chainIntent is an external api so its a breaking change to rename fields later. Other more internal code is probably less difficult to rename later

agusduha replied:
Fixed https://github.com/ethereum-optimism/optimism/pull/17412/commits/4ad8a301948a09f77c2d239ef15604d76fd8546c

tynes commented:
Just rebased this on top of the latest develop (https://github.com/ethereum-optimism/optimism/pull/17406)
The main things that need to be addressed:
- Put a CGT feature [here](https://github.com/ethereum-optimism/optimism/blob/8749b77f4d6b4767e40d11371ac3d37cb7f2f2d8/.circleci/config.yml#L253)
- Merge `OptimismPortal2CGT.t.sol` into `OptimismPortal2.t.sol` and gate tests using a function like `skipIfSysFeatureDisabled(CGT)`
- Instead of updating the Input to the L2Genesis script, use a bitmap `bytes32` and have each bit correspond to a different feature, `op-deployer` can set the bits, this reduces churn with adding large diffs for each new feature

scharissis commented on op-chain-ops/genesis/config.go:
```diff
@@ -275,18 +275,27 @@ func (d *GasPriceOracleDeployConfig) OperatorFeeParams() [32]byte {
+			return fmt.Errorf("%w: GasPayingTokenName cannot be empty", ErrInvalidDeployConfig)
+		}
+		if d.GasPayingTokenSymbol == "" {
+			return fmt.Errorf("%w: GasPayingTokenSymbol cannot be empty", ErrInvalidDeployConfig)
+		}
+		if d.NativeAssetLiquidityAmount == nil {
+			return fmt.Errorf("%w: NativeAssetLiquidityAmount cannot be nil", ErrInvalidDeployConfig)
 		}
-		log.Info("Using custom gas token", "address", d.CustomGasTokenAddress)
```
This logger (input parameter) now seems unused.
Do we want to log any of these errors? If not, let's remove the unused `log log.Logger`?

agusduha replied:
Fixed https://github.com/ethereum-optimism/optimism/pull/17412/commits/dfe5282a958e2aeac3fce599e4a2f0cb174c6cb5

scharissis commented on op-deployer/pkg/deployer/pipeline/l2genesis.go:
```diff
@@ -165,5 +188,6 @@ func defaultOverrides() l2GenesisOverrides {
 		SequencerFeeVaultWithdrawalNetwork:       "local",
 		EnableGovernance:                         false,
 		GovernanceTokenOwner:                     standard.GovernanceTokenOwner,
+		NativeAssetLiquidityAmount:               (*hexutil.Big)(maxUint248),
```
Why max and not min? (min just feels like a safer default)

agusduha replied:
Changed default to zero for safety https://github.com/ethereum-optimism/optimism/pull/17412/commits/dfe5282a958e2aeac3fce599e4a2f0cb174c6cb5

scharissis commented on op-deployer/pkg/deployer/pipeline/l2genesis.go:
```diff
@@ -22,6 +22,9 @@ import (
 )
 
 type l2GenesisOverrides struct {
+	UseCustomGasToken                        bool                      `json:"useCustomGasToken"`
+	GasPayingTokenName                       string                    `json:"gasPayingTokenName"`
+	GasPayingTokenSymbol                     string                    `json:"gasPayingTokenSymbol"`
```
Not sure what the convention is here, but I would think it a good idea to document which set of configuration options pertain to CGT, for future reference.


```suggestion
type l2GenesisOverrides struct {
    // CGT
	UseCustomGasToken                        bool                      `json:"useCustomGasToken"`
	GasPayingTokenName                       string                    `json:"gasPayingTokenName"`
	GasPayingTokenSymbol                     string                    `json:"gasPayingTokenSymbol"`
```

agusduha replied:
Fixed https://github.com/ethereum-optimism/optimism/pull/17412/commits/dfe5282a958e2aeac3fce599e4a2f0cb174c6cb5

scharissis reviewed (commented):
It would be great if we could add acceptance tests for CGT (`op-acceptance-tests/tests`) as part of this PR.

scharissis commented:
I've implemented a starting set of acceptance tests, which we could merge into this branch:
https://github.com/ethereum-optimism/optimism/pull/17451

geoknee commented on packages/contracts-bedrock/src/L1/SystemConfig.sol:
```diff
@@ -517,4 +517,10 @@ contract SystemConfig is ProxyAdminOwnedBase, OwnableUpgradeable, Reinitializabl
     function guardian() public view returns (address) {
         return superchainConfig.guardian();
     }
+
+    /// @notice Returns whether the custom gas token feature is enabled.
+    /// @return bool True if the custom gas token feature is enabled, false otherwise.
+    function isCustomGasToken() public view returns (bool) {
+        return isFeatureEnabled[Features.CUSTOM_GAS_TOKEN];
+    }
```
Does this change imply that the semver should be bumped?

agusduha replied:
Fixed https://github.com/ethereum-optimism/optimism/pull/17412/commits/dfe5282a958e2aeac3fce599e4a2f0cb174c6cb5

hexshire commented on packages/contracts-bedrock/src/L1/OptimismPortal2.sol:
```diff
@@ -574,6 +585,10 @@ contract OptimismPortal2 is Initializable, ResourceMetering, ReinitializableBase
         payable
         metered(_gasLimit)
     {
+        if (_isUsingCustomGasToken()) {
+            if (msg.value > 0) revert OptimismPortal_NotAllowedOnCGTMode();
+        }
+
```
dismiss

ashitakah replied:
dismiss

ashitakah commented on packages/contracts-bedrock/src/L2/LiquidityController.sol:
```diff
@@ -0,0 +1,103 @@
+    /// @param _to The address to receive the minted native asset
+    /// @param _amount The amount of native asset to mint and send
+    function mint(address _to, uint256 _amount) external {
+        if (!minters[msg.sender]) revert Unauthorized();
```
dismiss

ashitakah commented on packages/contracts-bedrock/src/L2/L2ToL1MessagePasserCGT.sol:
```diff
@@ -0,0 +1,36 @@
+    /// @notice Sends a message from L2 to L1.
+    /// @param _target   Address to call on L1 execution.
+    /// @param _gasLimit Minimum gas limit for executing the message on L1.
+    /// @param _data     Data to forward to L1 target.
+    function initiateWithdrawal(address _target, uint256 _gasLimit, bytes memory _data) public payable override {
+        if (IL1Block(Predeploys.L1_BLOCK_ATTRIBUTES).isCustomGasToken() && msg.value > 0) {
+            revert L2ToL1MessagePasserCGT_NotAllowedOnCGTMode();
+        }
+        super.initiateWithdrawal(_target, _gasLimit, _data);
+    }
+}
```
dismiss

agusduha committed:
fix: custom gas token rebase (#17484)

* fix: import in OptimismPortal2CGT test and pre-pr

* fix(cgt): add missing native asset amount (#543)

* fix: upgrade contract  name

* feat: add nativeAssetLiquidityAmount to config json files

* fix: add correct nativeAssetLiquidityAmount in op-deployer

* fix: add correct nativeAssetLiquidityAmount in op-e2e and fix withCustomGasToken argument on op-devstack

* fix: restore OptimismPortal2CGT

* fix: all comments

* fix: comments

* fix: governace

* fix: remove aux

* fix: remove aux

* fix: deploy cgt

* fix: revert cgt deploy config

* fix: deploy config

* chore: run linter

* fix: remove unnecessary GetNativeAssetLiquidityAmount

---------

Co-authored-by: Ashitaka <ashitaka@defi.sucks>
Co-authored-by: agusduha <agusnduha@gmail.com>
Co-authored-by: hexshire <hex@wonderland.xyz>

* fix: l2 genesis pipeline (#554)

* fix: add nativeAssetLiquidityAmount in e2e apply test (#555)

* fix: failing test

---------

Co-authored-by: niha <205694301+0xniha@users.noreply.github.com>
Co-authored-by: Ashitaka <ashitaka@defi.sucks>
Co-authored-by: hexshire <hex@wonderland.xyz>
Co-authored-by: Hex <165055168+hexshire@users.noreply.github.com>

agusduha committed:
Merge branch 'develop' into sc-feat/custom-gas-token-rebase

agusduha committed:
fix: semver lock

smartcontracts commented on packages/contracts-bedrock/src/L1/OPContractsManager.sol:
```diff
@@ -1072,6 +1072,12 @@ contract OPContractsManagerDeployer is OPContractsManagerBase {
 
+        // If the custom gas token feature was requested, enable the custom gas token feature in the SystemConfig
+        // contract.
+        if (_input.useCustomGasToken) {
```
Although this works, it'd be nicer if we could configure system customizations like this in a more generic way (e.g., an array of string names, or a bitmap, or something similar). If we go with this pattern then we're going to have breaking changes to the input ABI every time we add a new thing. (useCustomGasToken, useEthLockbox, useCgt). Each change to the OPCM ABI creates a bunch of problems. We do a bitmap for dev features (`devFeatureBitmap`), op-deployer probably supports arrays well enough now that we can use an array here instead, so you'd supply something like: `_input.enabledSystemFeatures: ["CUSTOM_GAS_TOKEN", "ETH_LOCKBOX"]`.

agusduha replied:
Fixing this here https://github.com/defi-wonderland/optimism/pull/562

mds1 replied:
What is the workflow here, e.g. when does the diff from the defi-wonderland org get migrated to this org?

agusduha replied:
@mds1 Opening a new fixes PR soon

agusduha replied:
Fixed https://github.com/ethereum-optimism/optimism/pull/17412/commits/9bb50bd10b1966f45895241bf7c37d4545d94946

smartcontracts commented on packages/contracts-bedrock/src/L1/OPContractsManager.sol:
```diff
@@ -1284,6 +1290,21 @@ contract OPContractsManagerDeployer is OPContractsManagerBase {
         (IResourceMetering.ResourceConfig memory referenceResourceConfig, ISystemConfig.Addresses memory opChainAddrs) =
             defaultSystemConfigParams(_input, _output);
 
+        return systemConfigInitializerData(_input, _superchainConfig, referenceResourceConfig, opChainAddrs);
```
What's the reason for this diff?

agusduha replied:
Little internal func refactor, we can revert it too

agusduha replied:
We will leave this change cause we were having stack too deep error in some tests

smartcontracts commented on packages/contracts-bedrock/src/L1/OptimismPortal2.sol:
```diff
@@ -233,9 +236,6 @@ contract OptimismPortal2 is Initializable, ResourceMetering, ReinitializableBase
         anchorStateRegistry = _anchorStateRegistry;
 
-        // Assert that the lockbox state is valid.
-        _assertValidLockboxState();
```
Why is this deleted?

agusduha replied:
This may have been lost in the rebase, will add it back

agusduha replied:
Fixed https://github.com/ethereum-optimism/optimism/pull/17412/commits/9bb50bd10b1966f45895241bf7c37d4545d94946

smartcontracts commented on packages/contracts-bedrock/src/L1/OptimismPortal2.sol:
```diff
@@ -561,6 +572,10 @@ contract OptimismPortal2 is Initializable, ResourceMetering, ReinitializableBase
         payable
         metered(_gasLimit)
     {
+        if (_isUsingCustomGasToken()) {
```
Auditors would probably come up with something like "user withdrawals are bricked if CGT is enabled after a withdrawal is created!". So we probably want a comment somewhere that explains that chains are not supposed to enable CGT outside of the initial deployment and doing so is a bad idea.

agusduha replied:
Yes! We can add a disclaimer under the CGT functions, this is something that can be extended to any features flag in the system config too

agusduha replied:
Fixed https://github.com/ethereum-optimism/optimism/pull/17412/commits/9bb50bd10b1966f45895241bf7c37d4545d94946

smartcontracts commented on packages/contracts-bedrock/src/L1/SystemConfig.sol:
```diff
@@ -539,4 +539,10 @@ contract SystemConfig is ProxyAdminOwnedBase, OwnableUpgradeable, Reinitializabl
+
+    /// @notice Returns whether the custom gas token feature is enabled.
+    /// @return bool True if the custom gas token feature is enabled, false otherwise.
+    function isCustomGasToken() public view returns (bool) {
```
Feels unnecessary, is there a particular reason to have this?

agusduha replied:
The old CGT design had a function like this one, we opt to maintain it as a legacy function for tools that used it

mds1 replied:
We should add the `/// @custom:legacy` tag then and remove it in a future release

agusduha replied:
Fixed https://github.com/ethereum-optimism/optimism/pull/17412/commits/9bb50bd10b1966f45895241bf7c37d4545d94946

smartcontracts commented on packages/contracts-bedrock/src/L2/L1BlockCGT.sol:
```diff
@@ -0,0 +1,72 @@
+            isCustomGasToken() ? ILiquidityController(Predeploys.LIQUIDITY_CONTROLLER).gasPayingTokenSymbol() : "ETH";
+    }
+
+    /// @notice Set chain to use custom gas token (callable by depositor account)
```
Is there a particular reason why this isn't just enabled by default? Why do we need a setter + deposit txn?

agusduha replied:
We are maintaining the same pattern used in the `L1Block` setters, we choose not to add a constructor/initializer. This can be helpful for NUTs in future CGT chains that wants to migrate, and also even for the standard genesis.

It is something we discussed with @tynes, but im not finding the PR where we did it

0xniha replied:
Here's the discussion: https://github.com/defi-wonderland/optimism/pull/478#discussion_r2285578236

mds1 commented on packages/contracts-bedrock/src/L2/L1BlockCGT.sol:
```diff
@@ -0,0 +1,72 @@
+        }
+    }
+
+    /// @notice Returns the gas paying token, its decimals, name and symbol.
+    function gasPayingToken() public pure override returns (address, uint8) {
+        revert("L1BlockCGT: deprecated");
+    }
```
If this just reverts, can we delete it?

agusduha replied:
If we delete it then it will inherit the parent behavior. Unless we want to change the `L1Block`

mds1 replied:
Ahh that makes sense, didn't realize this was already in the base L1Block

ashitakah commented on packages/contracts-bedrock/scripts/deploy/DeployConfig.s.sol:
```diff
@@ -120,6 +123,10 @@ contract DeployConfig is Script {
+        useCustomGasToken = _readOr(_json, "$.useCustomGasToken", false);
+        gasPayingTokenName = _readOr(_json, "$.gasPayingTokenName", "");
+        gasPayingTokenSymbol = _readOr(_json, "$.gasPayingTokenSymbol", "");
+        nativeAssetLiquidityAmount = _readOr(_json, "$.nativeAssetLiquidityAmount", type(uint248).max);
```
dismiss

agusduha committed:
fix: CGT review fixes (#17534)

* fix: json default zero

* feat(cgt): add native asset liquidity amount test & omitempty on cgt struct

* test: check native asset liquidity amount is correctly configured

* feat: add omitempty to name and symbol on cgt intent struct

* fix: check value in cgt

* feat(cgt): add cgt dev feature

* feat: add cgt dev feature

* refactor: set custom gas token (#563)

* feat: add cgt devFeatures and remove useCustomGasToken from DeployOPChainInput

---------

Co-authored-by: Hex <165055168+hexshire@users.noreply.github.com>

* fix: ir informational

* fix: informational

* fix: legacy tag

* feat(cgt): add setNativeAssetLiquidityAmount & L2Genesis max amount check

* feat: add setNativeAssetLiquidityAmount & L2Genesis max amount check

* test: fix bound param in liquidity controller

* fix: remove unused import

* refactor: move max amount check inside setNativeAssetLiquidity

* refactor: make chainIntent.CustomGasToken a non-pointer (#574)

* refactor: make chainIntent.CustomGasToken a non-pointer

* test: fix custom gas token text (#576)

---------

Co-authored-by: Hex <165055168+hexshire@users.noreply.github.com>

* fix: undo tests

Co-authored-by: hexshire <hex@wonderland.xyz>

* fix: cgt portal (#577)

* fix: semver

* fix: cgt tests coverage (#582)

* fix: l2genesis expectRevert amount test

* test: improve l1block coverage

* test: improve l1blockcgt coverage

* test: add isCustomGasToken tests on SystemConfig

* test: improve L2ToL1MessagePasserCGT coverage

* test: improve L2ToL1MessagePasser coverage

* fix: naming pre-pr

* fix(cgt): change L2ToL1MessagePasser & OPContractsManager semver (#584)

* fix: change L2ToL1MessagePasser semver

* fix: change L2ToL1MessagePasser & OPContractsManager  semver

* fix: cgt feature tests (#585)

* fix: cgt feature tests

* fix: deposits with amount 0 when cgt enabled

* Revert "fix: deposits with amount 0 when cgt enabled"

This reverts commit 8bee464d45cf7d4040c4dd1459ffc78de2508024.

* fix: cgt portal (#587)

* fix: ci fixes (#588)

---------

Co-authored-by: Ashitaka <96790496+ashitakah@users.noreply.github.com>
Co-authored-by: niha <205694301+0xniha@users.noreply.github.com>
Co-authored-by: Hex <165055168+hexshire@users.noreply.github.com>
Co-authored-by: hexshire <hex@wonderland.xyz>

bitwiseguy commented on op-deployer/pkg/deployer/state/chain_intent.go:
```diff
@@ -119,9 +127,32 @@ func (c *ChainIntent) Check() error {
+
+		if c.CustomGasToken.NativeAssetLiquidityAmount == nil {
+			return fmt.Errorf("%w: CustomGasToken.NativeAssetLiquidityAmount must be set when custom gas token is enabled, chainId=%s", ErrIncompatibleValue, c.ID)
+		}
```
Should we also ensure this is a positive big.Int?

agusduha replied:
We decided to not make it a strict requirement since maybe there can be an use case where the chain wants to start with zero. Either way the liquidity can be funded later directly to the `NativeAssetLiquidity` contract through an upgrade tx

bitwiseguy replied:
Ok sounds like 0 value is acceptable. Should we ensure its not negative?

agusduha replied:
We are assuming that's not needed since the L2 genesis script will revert in that case

bitwiseguy replied:
This chainIntent is created before the L2 genesis script runs though. I think it would be a nice protection to add. Better to fail early than far downstream. This is the op-deployer flow:

1. `op-deployer init` - creates intent.toml
2. user populates intent.toml (which contains the chainIntent structs)
3. `op-deployer apply` - executes the following sequential pipeline
    1. `DeploySuperchain.s.sol` (if needed)
    2. `DeployImplementations.s.sol` (if needed) 
    3. `DeployOPChain.s.sol` 
    4. `L2Genesis.s.sol`

agusduha replied:
Fixed https://github.com/ethereum-optimism/optimism/pull/17412/commits/4ad8a301948a09f77c2d239ef15604d76fd8546c

agusduha committed:
Merge branch 'develop' into sc-feat/custom-gas-token-rebase

agusduha committed:
fix: semver

agusduha committed:
fix: opcm version

agusduha committed:
Merge branch 'develop' into sc-feat/custom-gas-token-rebase

agusduha committed:
fix: tests

scharissis committed:
feat(op-acceptance-tests): add acceptance tests for native CGT across L1/L2 (#17451)

- Gate suite via L1Block.isCustomGasToken(); assert name/symbol non-empty
- L2: value transfer pays CGT token (typed amounts via op-service/eth)
- L2: XDM rejects callvalue; L2StandardBridge legacy withdraw reverts
- L1: Portal receive/deposit rejects ETH; introspect SystemConfig addr
- L1: assert SystemConfig.isCustomGasToken() is true

tynes commented on op-acceptance-tests/tests/custom_gas_token/cgt_l1_portal_introspection_test.go:
```diff
@@ -0,0 +1,55 @@
+	defer cancel()
+
+	// Portal exposes systemConfig() -> address
+	const portalABI = `[
```
[w3](https://github.com/lmittmann/w3?tab=readme-ov-file#abi-bindings) is useful here when creating callers

scharissis replied:
Oh yeah, I forgot about this, thanks!

https://github.com/ethereum-optimism/optimism/pull/17583

smartcontracts commented on packages/contracts-bedrock/scripts/checks/test-validation/exclusions.toml:
```diff
@@ -56,6 +56,7 @@ contract_name_validation = [
     "test/L2/GasPriceOracle.t.sol",                       # Contains contracts not matching GasPriceOracle base name
     "test/universal/StandardBridge.t.sol",                # Contains contracts not matching StandardBridge base name
     "test/L1/OPContractsManagerContractsContainer.t.sol", # Contains contracts not matching OPContractsManagerContractsContainer base name
+    "test/L1/OptimismPortal2.t.sol",                      # Contains contracts not matching OptimismPortal2 base name
```
Let's exclude the specific failing tests, OptimismPortal is a huge test file so would prefer not to exclude the whole thing

0xniha replied:
Fixed [`4f01447`](https://github.com/ethereum-optimism/optimism/pull/17412/commits/4f014476f98873ed16de4843cfcb7f8839ccd83c)

smartcontracts commented on packages/contracts-bedrock/src/L2/L1BlockCGT.sol:
```diff
@@ -0,0 +1,72 @@
+
+    /// @notice Set chain to use custom gas token (callable by depositor account)
+    function setCustomGasToken() external {
+        require(
```
Where does this get called? I don't see any diff to `op-node` in this PR that would cause the depositor account to actually trigger this code but maybe I'm missing something about how that works

smartcontracts replied:
Blocks audit

tynes replied:
This was introduced as part of the upgrade code path but was descoped, so we can remove this now and reintroduce it when we implement the upgrade (network upgrade tx meant to call this to migrate legacy cgt chains to the new implementation)

agusduha replied:
This is being used in the L2 genesis too, it is the only way we have to activate the CGT flag in L2. If we need to remove it, then we will need to add a `initialize` function to the `L1Block`

agusduha replied:
We will leave it to activate the CGT flag in L2

smartcontracts commented on packages/contracts-bedrock/src/L2/L2ToL1MessagePasserCGT.sol:
```diff
@@ -0,0 +1,36 @@
+    /// @notice The error thrown when a withdrawal is initiated with value and custom gas token is used.
+    error L2ToL1MessagePasserCGT_NotAllowedOnCGTMode();
+
+    /// @custom:semver +custom-gas-token.1
```
Nit: in L1BlockCGT this is just `+custom-gas-token`

0xniha replied:
Fixed [`4f01447`](https://github.com/ethereum-optimism/optimism/pull/17412/commits/4f014476f98873ed16de4843cfcb7f8839ccd83c)

smartcontracts commented on packages/contracts-bedrock/src/L2/LiquidityController.sol:
```diff
@@ -0,0 +1,103 @@
+    /// @notice Authorizes an address to perform liquidity control operations
+    /// @param _minter The address to authorize as a minter
+    function authorizeMinter(address _minter) external {
+        if (msg.sender != IProxyAdmin(Predeploys.PROXY_ADMIN).owner()) revert Unauthorized();
```
The L2 Proxy Admin Owner is the aliased L1 Proxy Admin Owner whenever Optimism Governance is managing upgrades for a chain. This would mean that the L1PAO is responsible for managing this auth. Are we confident this is the auth that we want? It may make sense for this to be a dedicated role instead.

smartcontracts replied:
Same comment everywhere else in this file where L2PAO is used

smartcontracts replied:
Blocks audit

tynes replied:
There is no strong reason to add another role to make the config more complex. If there was an obvious reason then I would be ok with adding a new role

mds1 replied:
I think the main question here is around the fact that if this role is the PAO, for chains using optimism governance, the security council needs to be involved every time we want to authorize or unauthorize minters — is that ok? Does it mean these calls need gov proposals or is some standard rollup charter change sufficient to avoid that? How often do we expect these methods to be called?

smartcontracts commented on packages/contracts-bedrock/src/L2/NativeAssetLiquidity.sol:
```diff
@@ -0,0 +1,57 @@
+/// @custom:predeploy 0x4200000000000000000000000000000000000029
+/// @title NativeAssetLiquidity
+/// @notice The NativeAssetLiquidity contract allows other contracts to access native asset liquidity
+contract NativeAssetLiquidity is ISemver {
```
Why are NativeAssetLiquidity and LiquidityController different contracts?

smartcontracts replied:
Blocks audit

tynes replied:
This was already covered in the design review. It was explicitly separated to reduce the risks of a bug, the less code in the contract with the liquidity the better

smartcontracts replied:
Ok, I will resolve this since it's late to be objecting to it.

smartcontracts commented on packages/contracts-bedrock/src/L2/NativeAssetLiquidity.sol:
```diff
@@ -0,0 +1,57 @@
+
+    /// @notice Fund the contract by sending native asset.
+    /// @dev The function is payable to accept native asset.
+    function fund() external payable {
```
What is the usecase for `fund`?

smartcontracts replied:
Blocks audit

tynes replied:
This can be removed, was part of the upgrade path

agusduha replied:
We are following the same pattern we did for the `ETHLiquidity`, we can remove it if needed

smartcontracts replied:
Would prefer to remove if there's no clear case for it yet. Worst case if we really need to send ETH to this contract we could always selfdestruct to it. Just want to avoid as much unnecessary code as possible since this contract is critical.

0xniha replied:
Removed [`4f01447`](https://github.com/ethereum-optimism/optimism/pull/17412/commits/4f014476f98873ed16de4843cfcb7f8839ccd83c)

smartcontracts commented on packages/contracts-bedrock/test/L1/OptimismPortal2.t.sol:
```diff
@@ -49,8 +49,8 @@ contract OptimismPortal2_TestInit is DisputeGameFactory_TestInit {
-    // Use a constructor to set the storage vars above, so as to minimize the number of ffi calls.
-    constructor() {
+    /// @dev Setup the system for a ready-to-use state.
+    function setUp() public virtual override {
```
I think this diff can be reverted

0xniha replied:
Fixed [`4f01447`](https://github.com/ethereum-optimism/optimism/pull/17412/commits/4f014476f98873ed16de4843cfcb7f8839ccd83c)

smartcontracts commented on packages/contracts-bedrock/test/L1/OptimismPortal2.t.sol:
```diff
@@ -73,10 +77,7 @@ contract OptimismPortal2_TestInit is DisputeGameFactory_TestInit {
-    }
 
-    /// @dev Setup the system for a ready-to-use state.
-    function setUp() public virtual override {
```
Can revert this part of the diff too, the only reason for the constructor diff was because of an older way the tests were being written

0xniha replied:
Fixed [`4f01447`](https://github.com/ethereum-optimism/optimism/pull/17412/commits/4f014476f98873ed16de4843cfcb7f8839ccd83c)

smartcontracts commented on packages/contracts-bedrock/test/L1/OptimismPortal2.t.sol:
```diff
@@ -694,6 +714,21 @@ contract OptimismPortal2_Receive_Test is OptimismPortal2_TestInit {
+        address portal = address(optimismPortal2);
+
+        vm.prank(alice);
+        vm.expectRevert(IOptimismPortal.OptimismPortal_NotAllowedOnCGTMode.selector);
```
```
        vm.expectRevert(bytes("StandardBridge: function can only be called from an EOA"));
        (bool revertsAsExpected,) = address(l1StandardBridge).call{ value: 100 }(hex"");
        assertTrue(revertsAsExpected, "expectRevert: call did not revert");
```

^ nit, this pattern from L1StandardBridge.t.sol is a bit simpler

0xniha replied:
Refactored [`4f01447`](https://github.com/ethereum-optimism/optimism/pull/17412/commits/4f014476f98873ed16de4843cfcb7f8839ccd83c)

smartcontracts commented on packages/contracts-bedrock/test/L1/OptimismPortal2.t.sol:
```diff
@@ -2479,9 +2565,18 @@ contract OptimismPortal2_DepositTransaction_Test is OptimismPortal2_TestInit {
     {
         // Prevent overflow on an upgrade context
-        _mint = bound(_mint, 0, type(uint256).max - address(ethLockbox).balance);
+        if (isUsingLockbox()) {
```
Did we run into a fuzz failure because this if statement wasn't here before? Just curious

hexshire replied:
We found and issue on this test, we are opening a PR to fix it.

0xniha replied:
Fixed `_mint` bounding: [`f033db8`](https://github.com/ethereum-optimism/optimism/pull/17412/commits/f033db8b83773c36a74a842c134d98d7b3f70843)

smartcontracts commented on packages/contracts-bedrock/test/L1/OptimismPortal2.t.sol:
```diff
@@ -2603,9 +2703,13 @@ contract OptimismPortal2_DepositTransaction_Test is OptimismPortal2_TestInit {
         bytes memory _data
     )
         external
+        virtual
```
Noticed there's a lot of `virtual` diff here - is this still necessary or is it a holdover from the previous Portal tests?

0xniha replied:
Fixed [`4f01447`](https://github.com/ethereum-optimism/optimism/pull/17412/commits/4f014476f98873ed16de4843cfcb7f8839ccd83c)

smartcontracts commented on packages/contracts-bedrock/test/L2/L1BlockCGT.t.sol:
```diff
@@ -0,0 +1,156 @@
+/// @title L1BlockCGT_SetL1BlockValues_Test
+/// @notice Tests the `setL1BlockValues` function of the `L1Block` contract with custom gas token
+///         enabled.
+contract L1BlockCGT_SetL1BlockValues_Test is L1Block_SetL1BlockValues_Test {
```
Since we're using the matrix option in CI, no longer necessary to have these inherited versions of the tests

0xniha replied:
Migrated tests from `L1BlockCGT` to `L1Block`: [`4f01447`](https://github.com/ethereum-optimism/optimism/pull/17412/commits/4f014476f98873ed16de4843cfcb7f8839ccd83c)

smartcontracts commented on packages/contracts-bedrock/test/L2/L1BlockCGT.t.sol:
```diff
@@ -0,0 +1,156 @@
+
+/// @title L1BlockCGT_Version_Test
+/// @notice Test contract for L1BlockCGT `version` function.
+contract L1BlockCGT_Version_Test is L1BlockCGT_TestInit {
```
General note - there's an option to not have a dedicated L1BlockCGT test file and instead merge into the base file using `skipIfDevFeatureDisabled/Enabled` as in the portal tests. Doesn't block merging for audit though.

0xniha replied:
Migrated tests from `L1BlockCGT` to `L1Block`: [4f01447](https://github.com/ethereum-optimism/optimism/pull/17412/commits/4f014476f98873ed16de4843cfcb7f8839ccd83c)

smartcontracts commented on packages/contracts-bedrock/test/L2/L2ToL1MessagePasserCGT.t.sol:
```diff
@@ -0,0 +1,87 @@
+
+/// @title L2ToL1MessagePasserCGT_TestInit
+/// @notice Tests the `L2ToL1MessagePasser` contract with a custom gas token enabled.
+contract L2ToL1MessagePasserCGT_TestInit is CommonTest {
```
Same comment as with L1BlockCGT, can also merge into base test file with skipIfDisabled/Enabled

0xniha replied:
Migrated tests from `L2ToL1MessagePasserCGT` to `L2ToL1MessagePasser`: [4f01447](https://github.com/ethereum-optimism/optimism/pull/17412/commits/4f014476f98873ed16de4843cfcb7f8839ccd83c)

smartcontracts commented on packages/contracts-bedrock/test/L2/LiquidityController.t.sol:
```diff
@@ -0,0 +1,272 @@
+
+    /// @notice Tests that contract is set up correctly.
+    function test_setup_succeeds() public view {
+        assertEq(liquidityController.version(), "1.0.0");
```
Nit - should really be tested as a separate LiquidityController_Version_Test (same for the other two assertions)

0xniha replied:
Fixed [`4f01447`](https://github.com/ethereum-optimism/optimism/pull/17412/commits/4f014476f98873ed16de4843cfcb7f8839ccd83c)

smartcontracts commented on packages/contracts-bedrock/test/L2/LiquidityController.t.sol:
```diff
@@ -0,0 +1,272 @@
+    }
+
+    /// @notice Shared modifier to authorize a minter.
+    modifier isAuthorizedMinter(address _minter) {
```
Nit - prefer to have this be a helper function (`_authorizeMinter(_minter)`)

smartcontracts replied:
Also this reads as a requirement rather than a state-modifying thing

0xniha replied:
Refactored [`4f01447`](https://github.com/ethereum-optimism/optimism/pull/17412/commits/4f014476f98873ed16de4843cfcb7f8839ccd83c)

smartcontracts commented on packages/contracts-bedrock/test/L2/LiquidityController.t.sol:
```diff
@@ -0,0 +1,272 @@
+    /// @notice Tests that the deauthorizeMinter function can be called by the owner.
+    function testFuzz_deauthorizeMinter_fromOwner_succeeds(address _minter) public {
+        // Set minter to authorized
+        stdstore.target(address(liquidityController)).sig(liquidityController.minters.selector).with_key(_minter)
```
Nit - there are a few places where the helper function could be used but the checked write is used instead

0xniha replied:
Fixed: [`f033db8`](https://github.com/ethereum-optimism/optimism/pull/17412/commits/f033db8b83773c36a74a842c134d98d7b3f70843)

smartcontracts commented on packages/contracts-bedrock/test/L2/NativeAssetLiquidity.t.sol:
```diff
@@ -0,0 +1,184 @@
+    }
+
+    /// @notice Tests that contract is set up correctly.
+    function test_setup_succeeds() public view {
```
Same nit as above, prefer not to test anything in the test initializer contract

smartcontracts replied:
Also for version tests throughout the PR, please just check that it returns a non-empty string so we can avoid changing this test every time the version changes in the contract

smartcontracts replied:
Future note to self to have "is valid semver" check somewhere

0xniha replied:
Fixed [`4f01447`](https://github.com/ethereum-optimism/optimism/pull/17412/commits/4f014476f98873ed16de4843cfcb7f8839ccd83c)

smartcontracts commented on packages/contracts-bedrock/test/L2/NativeAssetLiquidity.t.sol:
```diff
@@ -0,0 +1,184 @@
+        // Call the withdraw function with insufficient balance
+        vm.prank(address(liquidityController));
+        // Expect revert with OutOfFunds
+        vm.expectRevert(bytes(""));
```
Nit - maybe consider an explicit custom error case for this? Thoughts?

0xniha replied:
Added a [`InsufficientBalance`](https://github.com/ethereum-optimism/optimism/blob/4f014476f98873ed16de4843cfcb7f8839ccd83c/packages/contracts-bedrock/src/L2/NativeAssetLiquidity.sol#L45) custom error in `NativeAssetLiquidity`:  [`4f01447`](https://github.com/ethereum-optimism/optimism/pull/17412/commits/4f014476f98873ed16de4843cfcb7f8839ccd83c)

smartcontracts commented on packages/contracts-bedrock/test/libraries/Predeploys.t.sol:
```diff
@@ -150,12 +157,28 @@ contract Predeploys_UnclassifiedInterop_Test is Predeploys_TestInit {
+
+    /// @notice Tests that the predeploy addresses are set correctly. They have code and the
+    ///         proxied accounts have the correct admin. Using custom gas token.
+    function test_predeploysWithCustomGasToken_succeeds() external {
```
Nit - can be included in the original test contract with `skipIfDevFeatureDisabled`

smartcontracts commented on packages/contracts-bedrock/test/scripts/L2Genesis.t.sol:
```diff
@@ -139,4 +177,81 @@ contract L2Genesis_Run_Test is L2Genesis_TestInit {
+
+    /// @notice Tests that the run function reverts when CGT is enabled and the withdrawal network type of the FeeVaults
+    /// is L1.
+    function test_run_cgt_reverts() external setInputCGTEnabled {
```
Nit - could probably separate this into different tests

0xniha replied:
Fixed [`4f01447`](https://github.com/ethereum-optimism/optimism/pull/17412/commits/4f014476f98873ed16de4843cfcb7f8839ccd83c)

smartcontracts commented on packages/contracts-bedrock/test/scripts/L2Genesis.t.sol:
```diff
@@ -139,4 +177,81 @@ contract L2Genesis_Run_Test is L2Genesis_TestInit {
     }
+
+    /// @dev Modifier to set up the input for L2Genesis with CGT enabled.
+    modifier setInputCGTEnabled() {
```
Nit - probably cleaner to add a `setUp` function with a default input and then just have the CGT tests change the input. Also same comment as in other file about modifier vs helper function

0xniha replied:
Fixed [`4f01447`](https://github.com/ethereum-optimism/optimism/pull/17412/commits/4f014476f98873ed16de4843cfcb7f8839ccd83c)

smartcontracts commented on packages/contracts-bedrock/test/setup/CommonTest.sol:
```diff
@@ -208,4 +216,9 @@ contract CommonTest is Test, Setup, Events {
         useUpgradedFork = false;
     }
+
+    function enableCustomGasToken() public {
```
`enableCustomGasToken` and `DEV_FEATURE__CUSTOM_GAS_TOKEN` are currently independently controlling the L2 and L1 sides, respectively. IMO it would be best to enforce that these things are either controlled by the same config value or that both config values are consistent. I'm ok with either approach. Right now though it would be possible for you to accidentally enable the L2 genesis side without enabling the L1 side, or vice versa, which would create confusing test failures.

smartcontracts replied:
Doesn't block audit but will want to resolve quickly

0xniha replied:
Fixed [`4f01447`](https://github.com/ethereum-optimism/optimism/pull/17412/commits/4f014476f98873ed16de4843cfcb7f8839ccd83c)

agusduha committed:
fix: custom gas token rebase review comments (#17577)

* chore: use w3 library

* fix: msg error

* refactor(cgt): rename amount & add geq 0 check (#600)

* feat: add NativeAssetLiquidityAmount sign check

* refactor: rename NativeAssetLiquidityAmount to InitialLiquidity

* refactor(cgt): rename json nativeAssetLiquidityAmount to initialLiquidity (#601)

* refactor: rename json nativeAssetLiquidityAmount to initialLiquidity

* fix: only add initialLiquidity to cgt intent struct

---------

Co-authored-by: Hex <165055168+hexshire@users.noreply.github.com>
Co-authored-by: Ashitaka <96790496+ashitakah@users.noreply.github.com>
Co-authored-by: niha <205694301+0xniha@users.noreply.github.com>

agusduha committed:
fix: cgt review (#17612)

* refactor: replace enableCustomGasToken with cgt devfeature

* feat: remove fund in NativeAssetLiquidity (#605)

* fix(cgt): remove OptimismPortal2 test from exclusions (#606)

* fix: remove OptimismPortal2 test from exclusions

* fix: remove virtual and restore constructor in portal test

* refactor: use low level call to receive test

* refactor(cgt): migrate L1Block and L2ToL1MessagePasser cgt tests (#608)

* refactor: migrate L1Block cgt tests

* refactor: migrate L2ToL1MessagePasser cgt tests

* fix: comments (#607)

* fix: comments

* fix: custom error

* fix: errors and tests

* fix: errors and tests

* fix: cgt tests (#610)

* fix: comments

* fix: custom error

* fix: errors and tests

* fix: errors and tests

* fix: tests

* fix: tests

* fix: weth

* fix: cgt tests (#611)

* fix: comments

* fix: custom error

* fix: errors and tests

* fix: errors and tests

* fix: tests

* fix: tests

* fix: weth

* fix: remove types

* fix: coverage tests (#612)

* fix: separate setUp function tests (#613)

---------

Co-authored-by: niha <205694301+0xniha@users.noreply.github.com>
Co-authored-by: Ashitaka <96790496+ashitakah@users.noreply.github.com>

simon-something committed:
test: custom gas token invariants (#17489)

* test(inv): setup and total sup inv

* test(inv): accounting invariants

* chore: doc

* fix: import in OptimismPortal2CGT test and pre-pr

* fix(cgt): add missing native asset amount (#543)

* fix: upgrade contract  name

* feat: add nativeAssetLiquidityAmount to config json files

* fix: add correct nativeAssetLiquidityAmount in op-deployer

* fix: add correct nativeAssetLiquidityAmount in op-e2e and fix withCustomGasToken argument on op-devstack

* fix: restore OptimismPortal2CGT

* fix: all comments

* fix: comments

* fix: governace

* fix: remove aux

* fix: remove aux

* fix: deploy cgt

* fix: revert cgt deploy config

* fix: deploy config

* chore: run linter

* fix: remove unnecessary GetNativeAssetLiquidityAmount

---------

Co-authored-by: Ashitaka <ashitaka@defi.sucks>
Co-authored-by: agusduha <agusnduha@gmail.com>
Co-authored-by: hexshire <hex@wonderland.xyz>

* fix: l2 genesis pipeline (#554)

* fix: add nativeAssetLiquidityAmount in e2e apply test (#555)

* fix: failing test

* chore: doc

Co-authored-by: graphite-app[bot] <96075541+graphite-app[bot]@users.noreply.github.com>

* chore: remove unused imports

* chore: fix semgrep

* chore: change error name

---------

Co-authored-by: drgorillamd <83670532+drgorillamd@users.noreply.github.com>
Co-authored-by: niha <205694301+0xniha@users.noreply.github.com>
Co-authored-by: Ashitaka <ashitaka@defi.sucks>
Co-authored-by: agusduha <agusnduha@gmail.com>
Co-authored-by: hexshire <hex@wonderland.xyz>
Co-authored-by: AgusDuha <81362284+agusduha@users.noreply.github.com>
Co-authored-by: Hex <165055168+hexshire@users.noreply.github.com>
Co-authored-by: graphite-app[bot] <96075541+graphite-app[bot]@users.noreply.github.com>

agusduha committed:
refactor: use skipIfDevFeatureDisabled for cgt predeploys test (#614) (#17621)

Co-authored-by: niha <205694301+0xniha@users.noreply.github.com>

agusduha committed:
Merge branch 'develop' into sc-feat/custom-gas-token-rebase

agusduha committed:
fix: tests

0xniha committed:
fix: auth minter helper & bound _mint in depositTransaction tests (#17642)

* refactor: auth minter helper tests (#618)

* fix: bound _mint correctly in depositTransaction test function (#619)

0xniha commented on packages/contracts-bedrock/src/L2/LiquidityController.sol:
```diff
@@ -0,0 +1,105 @@
+    }
+
+    /// @notice Mints native asset liquidity and sends it to a specified address
+    /// @param _to The address to receive the minted native asset
```
dismiss

opgitgovernance commented:
This pr has been automatically marked as stale and will be closed in 5 days if no updates

drinkcoffee commented:
@tynes , should this PR be still open?

agusduha commented:
This PR follows in https://github.com/ethereum-optimism/optimism/pull/18076
